File: java/com/google/copybara/config/SkylarkParser.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.config;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.copybara.exception.ValidationException.checkCondition;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Maps;
import com.google.common.flogger.FluentLogger;
import com.google.copybara.ModuleSet;
import com.google.copybara.doc.annotations.Library;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.util.console.Console;
import com.google.copybara.util.console.StarlarkMode;
import java.io.IOException;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.function.Supplier;
import net.starlark.java.annot.StarlarkAnnotations;
import net.starlark.java.annot.StarlarkBuiltin;
import net.starlark.java.eval.EvalException;
import net.starlark.java.eval.Module;
import net.starlark.java.eval.Mutability;
import net.starlark.java.eval.Starlark;
import net.starlark.java.eval.StarlarkSemantics;
import net.starlark.java.eval.StarlarkThread;
import net.starlark.java.syntax.FileOptions;
import net.starlark.java.syntax.ParserInput;
import net.starlark.java.syntax.Program;
import net.starlark.java.syntax.StarlarkFile;
import net.starlark.java.syntax.SyntaxError;

/**
 * Loads Copybara configs out of Skylark files.
 */
public class SkylarkParser {

  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
  private static final String BARA_SKY = ".bara.sky";
  // For now all the modules are namespaces. We don't use variables except for 'core'.
  private final Iterable<Class<?>> modules;
  private final StarlarkMode validation;

  public SkylarkParser(Set<Class<?>> staticModules, StarlarkMode validation) {
    this.modules = ImmutableSet.<Class<?>>builder()
        .add(GlobalMigrations.class)
        .addAll(staticModules).build();
    this.validation = validation;
  }

  public Config loadConfig(ConfigFile config, ModuleSet moduleSet, Console console)
      throws IOException, ValidationException {
    return getConfigWithTransitiveImports(config, moduleSet, console).config;
  }

  private Config loadConfigInternal(ConfigFile content, ModuleSet moduleSet,
      Supplier<ImmutableMap<String, ConfigFile>> configFilesSupplier, Console console)
      throws IOException, ValidationException {
    GlobalMigrations globalMigrations;
    Module module;
    try {
      module = new Evaluator(moduleSet, content, configFilesSupplier, console).eval(content);
      globalMigrations = GlobalMigrations.getGlobalMigrations(module);
    } catch (InterruptedException e) {
      // This should not happen since we shouldn't have anything interruptable during loading.
      throw new RuntimeException("Internal error", e);
    }
    return new Config(
        globalMigrations.getMigrations(), content.path(), module.getPredeclaredBindings());
  }

  @VisibleForTesting
  public Module executeSkylark(ConfigFile content, ModuleSet moduleSet, Console console)
      throws IOException, ValidationException, InterruptedException {
    CapturingConfigFile capturingConfigFile = new CapturingConfigFile(content);
    ConfigFilesSupplier configFilesSupplier = new ConfigFilesSupplier();

    Module module = new Evaluator(moduleSet, content, configFilesSupplier, console).eval(content);
    configFilesSupplier.setConfigFiles(capturingConfigFile.getAllLoadedFiles());
    return module;
  }

  /**
   * Collect all ConfigFiles retrieved by the parser while loading {code config}.
   *
   * @param config Root file of the configuration.
   * @param console the console to use for printing error/information
   * @return A map linking paths to the captured ConfigFiles and the parsed Config
   * @throws IOException If files cannot be read
   * @throws ValidationException If config is invalid, references an invalid file or contains
   *     dependency cycles.
   */
  public ConfigWithDependencies getConfigWithTransitiveImports(
      ConfigFile config, ModuleSet moduleSet, Console console)
      throws IOException, ValidationException {
    CapturingConfigFile capturingConfigFile = new CapturingConfigFile(config);
    ConfigFilesSupplier configFilesSupplier = new ConfigFilesSupplier();

    Config parsedConfig = loadConfigInternal(capturingConfigFile, moduleSet, configFilesSupplier,
        console);

    ImmutableMap<String, ConfigFile> allLoadedFiles = capturingConfigFile.getAllLoadedFiles();

    configFilesSupplier.setConfigFiles(allLoadedFiles);

    return new ConfigWithDependencies(allLoadedFiles, parsedConfig);
  }

  private static class ConfigFilesSupplier
      implements Supplier<ImmutableMap<String, ConfigFile>> {

    private ImmutableMap<String, ConfigFile> configFiles = null;

    void setConfigFiles(ImmutableMap<String, ConfigFile> configFiles) {
      Preconditions.checkState(this.configFiles == null, "Already set");
      this.configFiles = checkNotNull(configFiles);
    }

    @Override
    public ImmutableMap<String, ConfigFile> get() {
      // We need to load all the files before knowing the set of files in the config.
      checkNotNull(configFiles, "Don't call the supplier before loading"
          + " finishes.");
      return configFiles;
    }
  }

  /**
   * A class that contains a loaded config and all the config files that were
   * accessed during the parsing.
   */
  public static class ConfigWithDependencies {
    private final ImmutableMap<String, ConfigFile> files;
    private final Config config;

    private ConfigWithDependencies(ImmutableMap<String, ConfigFile> files, Config config) {
      this.config = config;
      this.files = files;
    }

    public Config getConfig() {
      return config;
    }

    public ImmutableMap<String, ConfigFile> getFiles() {
      return files;
    }
  }

  /**
   * An utility class for traversing and evaluating the config file dependency graph.
   */
  private final class Evaluator {

    private final LinkedHashSet<String> pending = new LinkedHashSet<>();
    private final Map<String, Module> loaded = new HashMap<>();
    private final Console console;
    private final ConfigFile mainConfigFile;
    // Predeclared environment shared by all files (modules) loaded.
    private final ImmutableMap<String, Object> environment;
    private final ModuleSet moduleSet;

    private Evaluator(ModuleSet moduleSet, ConfigFile mainConfigFile,
        Supplier<ImmutableMap<String, ConfigFile>> configFilesSupplier,
        Console console) {
      this.console = checkNotNull(console);
      this.mainConfigFile = checkNotNull(mainConfigFile);
      this.moduleSet = checkNotNull(moduleSet);
      this.environment = createEnvironment(this.moduleSet, configFilesSupplier);
    }

    private Module eval(ConfigFile content)
        throws IOException, ValidationException, InterruptedException {
      if (pending.contains(content.path())) {
        throw throwCycleError(content.path());
      }
      Module module = loaded.get(content.path());
      if (module != null) {
        return module;
      }
      pending.add(content.path());

      // Make the modules available as predeclared bindings.
      // For modules that implement OptionsAwareModule, options are set in the object so
      // that the module can construct objects that require options.
      StarlarkSemantics semantics = StarlarkSemantics.DEFAULT;
      module = Module.withPredeclared(semantics, environment);

      // parse & compile
      ParserInput input = ParserInput.fromUTF8(content.readContentBytes(), content.path());
      FileOptions options =
          FileOptions.DEFAULT.toBuilder()
              // Ordinarily, load statements should create file-local variables.
              // For now, we make them create first-class members of Module.globals.
              .loadBindsGlobally(true)
              .allowToplevelRebinding(true) // allow e.g. x=1; x=2 at top level
              .requireLoadStatementsFirst(validation == StarlarkMode.STRICT)
              .build();

      Program prog;
      try {
        prog = Program.compileFile(StarlarkFile.parse(input, options), module);
      } catch (SyntaxError.Exception ex) {
        for (SyntaxError error : ex.errors()) {
          console.error(error.toString());
        }
        checkCondition(false, "Error loading config file.");
        return null; // unreachable
      }

      // process loads
      Map<String, Module> loadedModules = new HashMap<>();
      for (String load : prog.getLoads()) {
        Module loadedModule = eval(content.resolve(load + BARA_SKY));
        loadedModules.put(load, loadedModule);
      }

      // execute
      updateEnvironmentForConfigFile(
          this::starlarkPrint, content, mainConfigFile, environment, moduleSet);
      try (Mutability mu = Mutability.create("CopybaraModules")) {
        StarlarkThread thread = new StarlarkThread(mu, semantics);
        thread.setLoader(loadedModules::get);
        thread.setPrintHandler(this::starlarkPrint);
        Starlark.execFileProgram(prog, module, thread);
      } catch (EvalException ex) {
        console.error(ex.getMessageWithStack());
        throw new ValidationException("Error loading config file", ex);
      } catch (Starlark.UncheckedEvalException uex) {
        console.error(uex.toString());
        // rethrow the UEX because it has the starlark stacktrace.
        if (uex.getCause() != null) {
          throw new ValidationException(
              "Error loading config file: " + uex.getCause().getMessage(), uex);
        }
        throw new ValidationException("Error loading config file", uex);
      }

      pending.remove(content.path());
      loaded.put(content.path(), module);
      return module;
    }

    private void starlarkPrint(StarlarkThread thread, String msg) {
      console.verbose(thread.getCallerLocation() + ": " + msg);
    }

    private ValidationException throwCycleError(String cycleElement)
        throws ValidationException {
      StringBuilder sb = new StringBuilder();
      for (String element : pending) {
        sb.append(element.equals(cycleElement) ? "* " : "  ");
        sb.append(element).append("\n");
      }
      sb.append("* ").append(cycleElement).append("\n");
      console.error("Cycle was detected in the configuration: \n" + sb);
      throw new ValidationException("Cycle was detected");
    }
  }

  /** Updates the module globals with information about the current loaded config file. */
  // TODO(copybara-team): evaluate the cleaner approach of saving the varying parts in the
  // StarlarkThread.setThreadLocal and leaving the modules alone as nature intended.
  private void updateEnvironmentForConfigFile(
      StarlarkThread.PrintHandler printHandler,
      ConfigFile currentConfigFile,
      ConfigFile mainConfigFile,
      Map<String, Object> environment,
      ModuleSet moduleSet) {
    for (Object module : moduleSet.getModules().values()) {
      // We mutate the module per file loaded. Not ideal but it is the best we can do.
      if (module instanceof LabelsAwareModule) {
        LabelsAwareModule m = (LabelsAwareModule) module;
        m.setConfigFile(mainConfigFile, currentConfigFile);
        m.setPrintHandler(printHandler);
      }
    }
    for (Class<?> module : modules) {
      logger.atInfo().log("Creating variable for %s", module.getName());
      // We mutate the module per file loaded. Not ideal but it is the best we can do.
      if (LabelsAwareModule.class.isAssignableFrom(module)) {
        LabelsAwareModule m = (LabelsAwareModule) environment.get(getModuleName(module));
        m.setConfigFile(mainConfigFile, currentConfigFile);
        m.setPrintHandler(printHandler);
      }
    }
  }

  /**
   * Create the environment for all evaluations (will be shared between all the dependent files
   * loaded).
   */
  private ImmutableMap<String, Object> createEnvironment(
      ModuleSet moduleSet, Supplier<ImmutableMap<String, ConfigFile>> configFilesSupplier) {
    Map<String, Object> env = Maps.newHashMap();
    for (Entry<String, Object> module : moduleSet.getModules().entrySet()) {
      logger.atInfo().log("Creating variable for %s", module.getKey());
      if (module.getValue() instanceof LabelsAwareModule) {
        ((LabelsAwareModule) module.getValue()).setAllConfigResources(configFilesSupplier);
      }
      // Modules shouldn't use the same name
      env.put(module.getKey(), module.getValue());
    }

    for (Class<?> module : modules) {
      logger.atInfo().log("Creating variable for %s", module.getName());
      // Create the module object and associate it with the functions
      ImmutableMap.Builder<String, Object> envBuilder = ImmutableMap.builder();
      try {
        StarlarkBuiltin annot = StarlarkAnnotations.getStarlarkBuiltin(module);
        if (annot != null) {
          envBuilder.put(annot.name(), module.getConstructor().newInstance());
        } else if (module.isAnnotationPresent(Library.class)) {
          Starlark.addMethods(envBuilder, module.getConstructor().newInstance());
        }
      } catch (ReflectiveOperationException e) {
        throw new LinkageError(e.getMessage(), e);
      }
      env.putAll(envBuilder.build());

      // Add the options to the module that require them
      if (OptionsAwareModule.class.isAssignableFrom(module)) {
        ((OptionsAwareModule) env.get(getModuleName(module))).setOptions(moduleSet.getOptions());
      }
      if (LabelsAwareModule.class.isAssignableFrom(module)) {
        ((LabelsAwareModule) env.get(getModuleName(module)))
            .setAllConfigResources(configFilesSupplier);
      }
    }
    return ImmutableMap.copyOf(env);
  }

  private static String getModuleName(Class<?> cls) {
    return cls.getAnnotation(StarlarkBuiltin.class).name();
  }
}

File: src/Copybara.Core/Config/SkylarkParser.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using System.Reflection;
using Copybara.Common;
using Copybara.Exceptions;
using Copybara.Util.Console;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Starlark.Annot;
using Starlark.Eval;
using Starlark.Syntax;
using Console = Copybara.Util.Console.Console;
using StarlarkRt = Starlark.Eval.Starlark;
using Module = Starlark.Eval.Module;
using Tuple = Starlark.Eval.Tuple;
using FileOptions = Starlark.Syntax.FileOptions;

namespace Copybara.Config;

/// <summary>Loads Copybara configs out of Skylark files.</summary>
public class SkylarkParser
{
    private static readonly ILogger Logger = NullLogger.Instance;
    private const string DefaultExtension = ".bara.sky";

    private static readonly ImmutableHashSet<string> AllowedLoadExtensions =
        ImmutableHashSet.Create(DefaultExtension, ".scl");

    private static readonly object VisibilityFunc = new VisibilityFunction();

    // For now all the modules are namespaces. We don't use variables except for 'core'.
    private readonly IReadOnlyList<Type> _modules;
    private readonly StarlarkMode _validation;

    public SkylarkParser(IReadOnlySet<Type> staticModules, StarlarkMode validation)
    {
        var builder = ImmutableArray.CreateBuilder<Type>();
        builder.Add(typeof(GlobalMigrations));
        builder.AddRange(staticModules);
        _modules = builder.ToImmutable();
        _validation = validation;
    }

    /// <exception cref="ValidationException"/>
    public Config LoadConfig(ConfigFile config, ModuleSet moduleSet, Console console) =>
        GetConfigWithTransitiveImports(config, moduleSet, console).GetConfig();

    private Config LoadConfigInternal(
        ConfigFile content,
        ModuleSet moduleSet,
        Func<ImmutableDictionary<string, ConfigFile>> configFilesSupplier,
        Console console)
    {
        Module module = new Evaluator(this, moduleSet, content, configFilesSupplier, console)
            .Eval(content);
        GlobalMigrations globalMigrations = GlobalMigrations.GetGlobalMigrations(module);
        return new Config(
            globalMigrations.GetMigrations(), content.Path(), module.GetPredeclaredBindings());
    }

    public Module ExecuteSkylark(ConfigFile content, ModuleSet moduleSet, Console console)
    {
        var capturingConfigFile = new CapturingConfigFile(content);
        var configFilesSupplier = new ConfigFilesSupplier();

        Module module = new Evaluator(this, moduleSet, content, configFilesSupplier.Get, console)
            .Eval(content);
        configFilesSupplier.SetConfigFiles(capturingConfigFile.GetAllLoadedFiles());
        return module;
    }

    /// <summary>
    /// Collect all ConfigFiles retrieved by the parser while loading <paramref name="config"/>.
    /// </summary>
    /// <param name="config">Root file of the configuration.</param>
    /// <param name="moduleSet">the module set providing the Starlark globals.</param>
    /// <param name="console">the console to use for printing error/information.</param>
    /// <returns>A map linking paths to the captured ConfigFiles and the parsed Config.</returns>
    /// <exception cref="ValidationException">If config is invalid, references an invalid file or
    /// contains dependency cycles.</exception>
    public ConfigWithDependencies GetConfigWithTransitiveImports(
        ConfigFile config, ModuleSet moduleSet, Console console)
    {
        var capturingConfigFile = new CapturingConfigFile(config);
        var configFilesSupplier = new ConfigFilesSupplier();

        Config parsedConfig =
            LoadConfigInternal(capturingConfigFile, moduleSet, configFilesSupplier.Get, console);

        ImmutableDictionary<string, ConfigFile> allLoadedFiles =
            capturingConfigFile.GetAllLoadedFiles();

        configFilesSupplier.SetConfigFiles(allLoadedFiles);

        return new ConfigWithDependencies(allLoadedFiles, parsedConfig);
    }

    private sealed class ConfigFilesSupplier
    {
        private ImmutableDictionary<string, ConfigFile>? _configFiles;

        internal void SetConfigFiles(ImmutableDictionary<string, ConfigFile> configFiles)
        {
            Preconditions.CheckState(_configFiles == null, "Already set");
            _configFiles = Preconditions.CheckNotNull(configFiles);
        }

        internal ImmutableDictionary<string, ConfigFile> Get()
        {
            // We need to load all the files before knowing the set of files in the config.
            return Preconditions.CheckNotNull(
                _configFiles, "Don't call the supplier before loading finishes.");
        }
    }

    /// <summary>A class that traverses and evaluates the config file dependency graph.</summary>
    private sealed class Evaluator
    {
        private readonly SkylarkParser _parser;
        private readonly HashSet<string> _pending = new();
        private readonly List<string> _pendingOrder = new();
        private readonly Dictionary<string, Module> _loaded = new();
        private readonly Console _console;
        private readonly ConfigFile _mainConfigFile;

        // Predeclared environment shared by all files (modules) loaded.
        private readonly ImmutableDictionary<string, object> _environment;
        private readonly ModuleSet _moduleSet;

        internal Evaluator(
            SkylarkParser parser,
            ModuleSet moduleSet,
            ConfigFile mainConfigFile,
            Func<ImmutableDictionary<string, ConfigFile>> configFilesSupplier,
            Console console)
        {
            _parser = parser;
            _console = Preconditions.CheckNotNull(console);
            _mainConfigFile = Preconditions.CheckNotNull(mainConfigFile);
            _moduleSet = Preconditions.CheckNotNull(moduleSet);
            _environment = _parser.CreateEnvironment(_moduleSet, configFilesSupplier);
        }

        internal Module Eval(ConfigFile content)
        {
            if (_pending.Contains(content.Path()))
            {
                throw ThrowCycleError(content.Path());
            }
            if (_loaded.TryGetValue(content.Path(), out Module? existing))
            {
                return existing;
            }
            _pending.Add(content.Path());
            _pendingOrder.Add(content.Path());

            // Make the modules available as predeclared bindings.
            StarlarkSemantics semantics = StarlarkSemantics.DEFAULT;
            Module module = Module.WithPredeclared(semantics, _environment);

            // parse & compile
            ParserInput input = ParserInput.FromUTF8(content.ReadContentBytes(), content.Path());
            FileOptions options =
                FileOptions.DEFAULT.ToBuilder()
                    // Ordinarily, load statements should create file-local variables.
                    // For now, we make them create first-class members of Module.globals.
                    .LoadBindsGlobally(true)
                    .AllowToplevelRebinding(true) // allow e.g. x=1; x=2 at top level
                    .RequireLoadStatementsFirst(_parser._validation == StarlarkMode.Strict)
                    .Build();

            Program prog;
            try
            {
                prog = Program.CompileFile(
                    StarlarkFile.Parse(input, options), StarlarkRt.ModuleAsResolverModule(module));
            }
            catch (SyntaxError.Exception ex)
            {
                foreach (SyntaxError error in ex.Errors)
                {
                    _console.Error(error.ToString());
                }
                throw new ValidationException("Error loading config file.");
            }

            // process loads
            var loadedModules = new Dictionary<string, Module>();
            var fileToLoad = new Dictionary<string, string>();
            foreach (string l in prog.GetLoads().Distinct())
            {
                string key = AllowedLoadExtensions.Any(l.EndsWith) ? l : l + DefaultExtension;
                fileToLoad[key] = l;
            }

            foreach (var entry in
                content
                    // Resolve all in one call so the implementor can do it in batch/parallel.
                    .ResolveAll(fileToLoad.Keys.ToImmutableHashSet()))
            {
                Module loadedModule = Eval(entry.Value);
                loadedModules[fileToLoad[entry.Key]] = loadedModule;
            }

            // execute
            _parser.UpdateEnvironmentForConfigFile(
                StarlarkPrint, content, _mainConfigFile, _environment, _moduleSet);
            using (Mutability mu = Mutability.Create("CopybaraModules"))
            {
                StarlarkThread thread = StarlarkThread.CreateTransient(mu, semantics);
                thread.SetLoader(m => loadedModules.TryGetValue(m, out Module? md) ? md : null);
                thread.SetPrintHandler(StarlarkPrint);
                try
                {
                    StarlarkRt.ExecFileProgram(prog, module, thread);
                }
                catch (EvalException ex)
                {
                    _console.Error(ex.Message);
                    throw new ValidationException("Error loading config file", ex);
                }
            }

            _pending.Remove(content.Path());
            _pendingOrder.Remove(content.Path());
            _loaded[content.Path()] = module;
            return module;
        }

        private void StarlarkPrint(StarlarkThread thread, string msg) =>
            _console.Verbose(thread.GetCallerLocation() + ": " + msg);

        private ValidationException ThrowCycleError(string cycleElement)
        {
            var sb = new System.Text.StringBuilder();
            foreach (string element in _pendingOrder)
            {
                sb.Append(element.Equals(cycleElement) ? "* " : "  ");
                sb.Append(element).Append('\n');
            }
            sb.Append("* ").Append(cycleElement).Append('\n');
            _console.Error("Cycle was detected in the configuration: \n" + sb);
            return new ValidationException("Cycle was detected");
        }
    }

    /// <summary>Updates the module globals with information about the current loaded config file.</summary>
    private void UpdateEnvironmentForConfigFile(
        StarlarkThread.PrintHandler printHandler,
        ConfigFile currentConfigFile,
        ConfigFile mainConfigFile,
        ImmutableDictionary<string, object> environment,
        ModuleSet moduleSet)
    {
        foreach (object module in moduleSet.GetModules().Values)
        {
            // We mutate the module per file loaded. Not ideal but it is the best we can do.
            if (module is ILabelsAwareModule m)
            {
                m.SetConfigFile(mainConfigFile, currentConfigFile);
                m.SetPrintHandler(printHandler);
            }
        }
        foreach (Type module in _modules)
        {
            Logger.LogInformation("Creating variable for {Module}", module.FullName);
            // We mutate the module per file loaded. Not ideal but it is the best we can do.
            if (typeof(ILabelsAwareModule).IsAssignableFrom(module))
            {
                var m = (ILabelsAwareModule)environment[GetModuleName(module)];
                m.SetConfigFile(mainConfigFile, currentConfigFile);
                m.SetPrintHandler(printHandler);
            }
        }
    }

    /// <summary>
    /// Create the environment for all evaluations (will be shared between all the dependent files
    /// loaded).
    /// </summary>
    private ImmutableDictionary<string, object> CreateEnvironment(
        ModuleSet moduleSet, Func<ImmutableDictionary<string, ConfigFile>> configFilesSupplier)
    {
        var env = new Dictionary<string, object>();
        foreach (var module in moduleSet.GetModules())
        {
            Logger.LogInformation("Creating variable for {Module}", module.Key);
            if (module.Value is ILabelsAwareModule lam)
            {
                lam.SetAllConfigResources(configFilesSupplier);
            }
            // Modules shouldn't use the same name
            env[module.Key] = module.Value;
        }

        foreach (Type module in _modules)
        {
            Logger.LogInformation("Creating variable for {Module}", module.FullName);
            // Create the module object and associate it with the functions
            var envBuilder = new Dictionary<string, object>();
            var annot = module.GetCustomAttribute<StarlarkBuiltinAttribute>(inherit: false);
            if (annot != null)
            {
                envBuilder[annot.Name] = Activator.CreateInstance(module)!;
            }
            else if (IsLibrary(module))
            {
                // Reference-forward: upstream registers @Library modules' methods directly. The
                // doc.annotations.Library attribute is not yet ported; handled via IsLibrary probe.
                StarlarkRt.AddMethods(envBuilder, Activator.CreateInstance(module)!);
            }

            foreach (var e in envBuilder)
            {
                env[e.Key] = e.Value;
            }

            // Add the options to the module that require them
            if (typeof(IOptionsAwareModule).IsAssignableFrom(module))
            {
                ((IOptionsAwareModule)env[GetModuleName(module)]).SetOptions(moduleSet.GetOptions());
            }
            if (typeof(ILabelsAwareModule).IsAssignableFrom(module))
            {
                ((ILabelsAwareModule)env[GetModuleName(module)])
                    .SetAllConfigResources(configFilesSupplier);
            }
        }
        env["visibility"] = VisibilityFunc;
        return env.ToImmutableDictionary();
    }

    private static string GetModuleName(Type cls) =>
        cls.GetCustomAttribute<StarlarkBuiltinAttribute>(inherit: false)!.Name;

    // Probes for a doc.annotations.Library-style attribute by name so the port keeps working once
    // that annotation lands, without a hard dependency on it here.
    private static bool IsLibrary(Type module) =>
        module.GetCustomAttributes(inherit: false)
            .Any(a => a.GetType().Name is "LibraryAttribute" or "Library");

    private sealed class VisibilityFunction : IStarlarkCallable
    {
        public object Call(StarlarkThread thread, Tuple args, Dict kwargs) => StarlarkRt.None;

        public object? Fastcall(StarlarkThread thread, object?[] positional, object?[] named) =>
            StarlarkRt.None;

        public string Name => "visibility";

        public void Repr(Printer printer, StarlarkSemantics semantics) =>
            printer.Append("<built-in function visibility>");

        public Location Location => Location.BUILTIN;
    }
}
